Add the ability to fetch the source code from the requests API endpoint

mprahl opened the pull request:
This still needs documentation and tests written, but I'd like to get your opinion on this.

lcarva commented on cachito/workers/scm.py:
```diff
@@ -0,0 +1,201 @@
+        :rtype: str
+        """
+        if not self._archive_path:
+            directory = os.path.join(self.archives_dir, *self.repo_name.split('/'))
```
Is splitting necessary? Won't this produce the same thing? 
```python
os.path.join(self.archives_dir, self.repo_name)
```

mprahl replied:
It would on systems that use forward slashes in their paths. I tend to not assume that when writing code, but it's probably safe to assume we won't be supporting Windows :stuck_out_tongue:

lcarva replied:
Fair point :)

lcarva commented on cachito/workers/scm.py:
```diff
@@ -0,0 +1,201 @@
+        self._archive_path = None
+        self._repo_name = None
+        # Validate that the base directory can be retrieved
+        self.archives_dir
```
It sounds like you don't want `self.archives_dir` to be a property :)
Maybe just have a helper method that is called here to do the work?

mprahl replied:
I'll keep it as a property and move the validation to `validate_celery_config` so this line can be removed.

lcarva commented on cachito/workers/scm.py:
```diff
@@ -0,0 +1,201 @@
+            # Import this here to avoid a circular import
+            from cachito.workers.tasks import app as celery_app
+
+            if not celery_app.conf.get('cachito_archives_dir'):
```
Can we make this a required attribute in the config?  I'd rather see an error during app start time then something that happens later on. IOW, let's move this to `validate_celery_config`.

mprahl replied:
Is it safe to assume that all of the celery workers will need access to the archives directory? If so, I can make it mandatory.

lcarva commented on cachito/workers/scm.py:
```diff
@@ -0,0 +1,201 @@
+                log.error(
+                    'Cloning the git repo with "%s" failed with: %s', ', '.join(cmd),
+                    error_output)
+                raise CachitoError(error)
```
Maybe set `error` to indicate it failed to clone?

lcarva commented on cachito/workers/tasks.py:
```diff
@@ -15,3 +18,22 @@
+    log.info('Fetching the source from "%s" at reference "%s"', url, ref)
+    try:
+        # Default to Git for now
+        scm = Git.create_from_url(url, ref)
```
`Git(url, ref)` ?

mprahl replied:
Oops, I forgot to update this after I changed it.

mprahl commented:
@lcarva okay, all the feedback has been addressed and I've added unit tests. I've also started validating the provided refs, and renamed the database columns to be generic and not specific to git. I figure it's easier to change this now rather than later.

lcarva commented on cachito/workers/tasks.py:
```diff
@@ -1,17 +1,48 @@
 configure_celery(app)
+# Only validate the configuration when Celery is directly running this module. This allows
+# importers of this module (e.g. the Flask app) to not have Celery fully configured.
+if os.path.basename(sys.argv[0]).startswith('celery'):
```
This is pretty hackish :)
Can we use [on_init](https://docs.celeryproject.org/en/latest/reference/celery.html#celery.Celery.on_init) ?

mprahl replied:
I think `on_init` wouldn't solve the issue, but it looks like [worker_init signal](https://docs.celeryproject.org/en/master/userguide/signals.html?highlight=signals#worker-init) might work. I'll try that.

mprahl commented:
@lcarva this should be all set now. Thanks for the review!
